libavformat/rtcenc.c

```diff
@@ -0,0 +1,2226 @@
+} DTLSContext;
+
+static int openssl_gen_private_key(DTLSContext *ctx)
+{
+    int ret = 0;
+#if OPENSSL_VERSION_NUMBER < 0x30000000L /* OpenSSL 3.0 */
+    EC_GROUP *ecgroup = NULL;
+    EC_KEY* dtls_eckey = NULL;
+#else
+    const char *curve = "prime256v1";
+#endif
+    void *s1 = ctx->log_avcl;
+
+    /* Should use the curves in ClientHello.supported_groups, for example:
+     *      Supported Group: x25519 (0x001d)
+     *      Supported Group: secp256r1 (0x0017)
+     *      Supported Group: secp384r1 (0x0018)
+     * Note that secp256r1 in openssl is called NID_X9_62_prime256v1 or prime256v1 in string,
+     * not NID_secp256k1 or secp256k1 in string
+     */
+#if OPENSSL_VERSION_NUMBER < 0x30000000L /* OpenSSL 3.0 */
+    ctx->dtls_pkey = EVP_PKEY_new();
+    dtls_eckey = EC_KEY_new();
+    ecgroup = EC_GROUP_new_by_curve_name(NID_X9_62_prime256v1);
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L // v1.1.x
+    /* For openssl 1.0, we must set the group parameters, so that cert is ok. */
+    EC_GROUP_set_asn1_flag(ecgroup, OPENSSL_EC_NAMED_CURVE);
+#endif
+
+    if (EC_KEY_set_group(dtls_eckey, ecgroup) != 1) {
+        av_log(s1, AV_LOG_ERROR, "DTLS: EC_KEY_set_group failed\n");
+        ret = AVERROR(EINVAL);
+        goto end;
+    }
+
+    if (EC_KEY_generate_key(dtls_eckey) != 1) {
+        av_log(s1, AV_LOG_ERROR, "DTLS: EC_KEY_generate_key failed\n");
+        ret = AVERROR(EINVAL);
+        goto end;
+    }
+
+    if (EVP_PKEY_set1_EC_KEY(ctx->dtls_pkey, dtls_eckey) != 1) {
+        av_log(s1, AV_LOG_ERROR, "DTLS: EVP_PKEY_set1_EC_KEY failed\n");
+        ret = AVERROR(EINVAL);
+        goto end;
+    }
+#else
+    ctx->dtls_pkey = EVP_EC_gen(curve);
+    if (!ctx->dtls_pkey) {
+        av_log(s1, AV_LOG_ERROR, "DTLS: EVP_EC_gen curve=%s failed\n", curve);
+        ret = AVERROR(EINVAL);
+        goto end;
+    }
+#endif
+
+end:
+#if OPENSSL_VERSION_NUMBER < 0x30000000L /* OpenSSL 3.0 */
+    EC_KEY_free(dtls_eckey);
+    EC_GROUP_free(ecgroup);
+#endif
+    return ret;
+}
+
+/**
+ * Generate a self-signed certificate and private key for DTLS.
+ */
+static av_cold int dtls_context_init(DTLSContext *ctx)
+{
+    int ret = 0, serial, expire_day, i, n = 0;
+    AVBPrint fingerprint;
+    unsigned char md[EVP_MAX_MD_SIZE];
+    const char *aor = "ffmpeg.org", *curve = NULL;
+    X509_NAME* subject = NULL;
+    X509 *dtls_cert = NULL;
+    void *s1 = ctx->log_avcl;
+
+    /* To prevent a crash during cleanup, always initialize it. */
+    av_bprint_init(&fingerprint, 1, MAX_SDP_SIZE);
+
+    ctx->dtls_cert = dtls_cert = X509_new();
+    if (!dtls_cert) {
+        ret = AVERROR(ENOMEM);
+        goto end;
+    }
+
+    /* Generate a private key to ctx->dtls_pkey. */
+    if ((ret = openssl_gen_private_key(ctx)) < 0)
+        goto end;
+
+    /* Generate a self-signed certificate. */
+    subject = X509_NAME_new();
+    if (!subject) {
+        ret = AVERROR(ENOMEM);
+        goto end;
+    }
+
+    serial = (int)av_get_random_seed();
```

Comment:
Derek Buitenhuis:

```text
> +    serial = (int)av_get_random_seed();
```

For all uses of av_get_random_seed: Why are we using a seed and not a PRNG?